### PR TITLE
SC-13592: Fix bugs for miscalculating included exceptions

### PR DESCRIPTION
**Pull request by ksk5280**
This fixes a bug that was found in adding days over a span with a `true` exception

### CODE REVIEW

File: src/addBusinessDays/index.ts [hunk 1]
```diff
@@ -42,49 +42,64 @@ export default function addBusinessDays(
...
   const isExcepted = (date: Date): boolean | null => {
     if (options.exceptions) {
       const exception =
         Object.keys(options.exceptions).find((e) =>
           isSameDay(new Date(e), date)
         ) || ''
-      if (options.exceptions[exception] !== undefined) {
+
+      if (options.exceptions[exception] === true) {
+        accountedForExceptions -= 1
```
Comment (ksk5280) [inline src/addBusinessDays/index.ts]: if true it needs to subtract from `accountedForExceptions`

File: src/addBusinessDays/index.ts [hunk 2]
```diff
@@ -42,49 +42,64 @@ export default function addBusinessDays(
...
         accountedForExceptions += 1
       }
       return options.exceptions[exception]
     }
     return null
   }
-  const isNonWorkingDay = (date: Date) =>
-    isExcepted(date) === false || !businessDays.includes(date.getDay())
+
+  const isWorkingDay = (date: Date) => {
```
Comment (ksk5280) [inline src/addBusinessDays/index.ts]: I changed this to the positive form to avoid the double negative of `!isNonWorkingDay`

File: src/addBusinessDays/index.ts [hunk 3]
```diff
@@ -42,49 +42,64 @@ export default function addBusinessDays(
...
+    const hasException = isExcepted(date)
+    if (hasException === false) return false
+    if (hasException === true || businessDays.includes(date.getDay())) {
```
Comment (ksk5280) [inline src/addBusinessDays/index.ts]: previously, if the exception was true, this function would return false because it wasn't a business day

File: src/addBusinessDays/index.ts [hunk 4]
```diff
@@ -42,49 +42,64 @@ export default function addBusinessDays(
...
-  const startedOnNonWorkingDay = isNonWorkingDay(date)
-  const hours = date.getHours()
+
+  const startedOnNonWorkingDay = !isWorkingDay(initialDate)
+  const hours = initialDate.getHours()
   const sign = amount < 0 ? -1 : 1
   const fullWeeks = toInteger(amount / businessDays.length)
-  date.setDate(date.getDate() + fullWeeks * 7)
+
+  let newDate = initialDate
```
Comment (ksk5280) [inline src/addBusinessDays/index.ts]: I changed the variable names because I found it a little confusing for everything to be called `date`

File: src/addBusinessDays/index.ts [hunk 5]
```diff
@@ -118,29 +133,31 @@ export default function addBusinessDays(
           return businessDaysDelta
       }
     }, 0) - accountedForExceptions
+
   // Add or subtract days until we have applied all our exceptions
   while (dayChangesDueToExceptions !== 0) {
     const deltaSign = dayChangesDueToExceptions < 0 ? -1 : 1
-    if (businessDays.includes(date.getDay())) {
+    if (isWorkingDay(newDate)) {
```
Comment (ksk5280) [inline src/addBusinessDays/index.ts]: this needed to check for a working day instead of a business day, to include or exclude the exceptions correctly

File: src/addBusinessDays/test.ts [hunk 6]
```diff
@@ -103,11 +103,53 @@ describe('addBusinessDays', function () {
       assert.deepStrictEqual(result, new Date(2022, 0 /* Jan */, 21))
     })
 
-    it('can override business days with exceptions', function () {
+    it('can include business days with true exceptions', function () {
```
Comment (ksk5280) [inline src/addBusinessDays/test.ts]: I added more tests for true exceptions, which were failing previously.

It would be great to refactor these to it.each tests, but at this point I just want to get this work finished...

File: src/addBusinessDays/test.ts [hunk 7]
```diff
@@ -103,11 +103,53 @@ describe('addBusinessDays', function () {
...
+      const result = addBusinessDays(new Date(2022, 1 /* Feb */, 9), 5, {
+        businessDays: [1, 2, 3, 4, 5, 6],
+        exceptions: { '02/13/22': true },
+      })
+
+      // Then we expect to have Sunday 2/13 included as a working day
+      assert.deepStrictEqual(result, new Date(2022, 1 /* Feb */, 14))
+    })
+
+    it('can include business days with true exceptions and ending on the exception', function () {
```
Comment (chr1sjf0x) [inline src/addBusinessDays/test.ts]: question: Do we have any starting on an exception tests?

Comment (ksk5280) [reply src/addBusinessDays/test.ts]: I added a couple

File: src/addBusinessDays/index.ts [hunk 8]
```diff
@@ -118,29 +131,34 @@ export default function addBusinessDays(
...
     }
-    date.setDate(date.getDate() + deltaSign)
+    newDate.setDate(newDate.getDate() + deltaSign)
   }
 
   // If we land on a non-working date, we add days accordingly to land on the next business day
   const reduceIfNonWorkingDay = (date: Date) => {
-    if (isNonWorkingDay(date) && amount !== 0) {
+    if (!isWorkingDay(date) && amount !== 0) {
       const newSign = startedOnNonWorkingDay ? -sign : sign
```
Comment (ksk5280) [inline src/addBusinessDays/index.ts]: @chr1sjf0x do you have any insight into why the `newSign` depends on whether `startedOnNonWorkingDay` is true? I have a failing test that starts and ends on excluded exceptions and this `newSign` ends up changing dates in the wrong direction, but if I try to just use `sign` then I have other tests that fail.

File: src/addBusinessDays/test.ts [hunk 9]
```diff
@@ -118,27 +160,63 @@ describe('addBusinessDays', function () {
...
+      // When we start on the exception and add 6 business days
+      const result = addBusinessDays(new Date(2022, 1 /* Feb */, 4), 6, {
+        exceptions: { '02/04/22': false },
+      })
+
+      // Then we expect to have the disabled Friday ignored
+      assert.deepStrictEqual(result, new Date(2022, 1 /* Feb */, 14))
+    })
+
+    it('can correctly add days when starting and ending on a false exception', function () {
```
Comment (ksk5280) [inline src/addBusinessDays/test.ts]: @chr1sjf0x this is the test that is failing, which theoretically shouldn't happen because our date picker won't allow it... so is it even a useful test?

File: src/addBusinessDays/test.ts [hunk 10]
```diff
@@ -118,27 +160,76 @@ describe('addBusinessDays', function () {
...
+      // When we start on an exception, add business days, and end on an exception
+      const result = addBusinessDays(new Date(2022, 1 /* Feb */, 4), 5, {
+        exceptions: { '02/04/22': false, '02/11/22': false },
+      })
+
+      // Then we expect to have the disabled days ignored
+      assert.deepStrictEqual(result, new Date(2022, 1 /* Feb */, 14))
+    })
+
+    it('can correctly add days when starting and ending on a true exception', function () {
```
Comment (ksk5280) [inline src/addBusinessDays/test.ts]: This is another failing edge case.

File: src/addBusinessDays/index.ts (hunk 3 above)
Comment (chr1sjf0x) [reply src/addBusinessDays/index.ts:66]: In one of the cases that Khalid fixed the other day, the behaviour of using `sign` as is resulted in too few business days being added when starting on a non-working day. This switch got it working, but at the expense of your new cases 🤔

File: src/addBusinessDays/index.ts (hunk 8 above)
Comment (chr1sjf0x) [reply src/addBusinessDays/index.ts]: In one of the cases that Khalid fixed the other day, the behaviour of using sign as is resulted in too few business days being added when starting on a non-working day. This switch got it working, but at the expense of your new cases 🤔

Comment (ksk5280) [reply src/addBusinessDays/index.ts]: We figured out that this needed to be `startedOnNonBusinessDay` to fix a couple tests

File: src/addBusinessDays/test.ts (hunk 9 above)
Comment (ksk5280) [reply src/addBusinessDays/test.ts]: this passes now

File: src/addBusinessDays/test.ts (hunk 10 above)
Comment (ksk5280) [reply src/addBusinessDays/test.ts]: this also passes now

File: src/addBusinessDays/test.ts (hunk 6 above)
Comment (chr1sjf0x) [reply src/addBusinessDays/test.ts]: question: A lot of tests are added here, but not in `subBusinessDays`. How confident are we that all these cases will `just work` in the opposite direction?

File: src/addBusinessDays/index.ts [hunk 11]
```diff
@@ -42,49 +42,70 @@ export default function addBusinessDays(
...
+  let newDate = new Date(initialDate)
+  let fullWeeks = toInteger(amount / businessDays.length)
   // Get remaining days not part of a full week
   let restDays = Math.abs(amount % businessDays.length)
+
+  // Fixes edge case bug to account for false exceptions in the first week
+  if (options.exceptions && restDays === 0 && fullWeeks === 1) {
+    restDays = businessDays.length
+    fullWeeks = 0
+  }
```
Comment (chr1sjf0x) [inline src/addBusinessDays/index.ts]: thought: Let's discuss this when you are online. I'd like to understand the issue/rational further. I'm also a little concerned this might be pushing the bugs to the second week?

File: src/addBusinessDays/index.ts [hunk 12]
```diff
@@ -37,110 +36,63 @@ export default function addBusinessDays(
...
-        ) || ''
-      if (options.exceptions[exception] !== undefined) {
-        accountedForExceptions += 1
-      }
-      return options.exceptions[exception]
-    }
-    return null
+  // Throw an exception if businessDays includes a number greater than 6
+  if (businessDays?.filter((number) => number > 6).length > 0) {
+    throw new RangeError('business days must be between 0 and 6')
```
Comment (chr1sjf0x) [inline src/addBusinessDays/index.ts]: thought: I was going to point out that the use of an array here seems like a poor choice, since one could specify `[6, 6, 6]`... however, I noticed that the new implementation at least is no longer using `businessDays.length` in its calculations, and wouldn't be susceptible to duplicate number issues. So won't make a blocking request for change, but might be worth circling back to that in the associated tech debt ticket.

File: src/addBusinessDays/index.ts [hunk 13]
```diff
@@ -37,110 +36,63 @@ export default function addBusinessDays(
...
-    if (!isNonWorkingDay(date)) restDays -= 1
+  // returns true or false if the date has an exception, else null
+  const findException = (date: Date): boolean | null => {
+    if (options.exceptions) {
+      const exceptionDate =
+        Object.keys(exceptions).find((e) => isSameDay(new Date(e), date)) || ''
+      return exceptions[exceptionDate]
+    }
+    return null
   }
```
Comment (chr1sjf0x) [inline src/addBusinessDays/index.ts]: request change: This feels like an inefficient way to find the exception for a given day (I would expect, at least if the number of exceptions is large). The exceptions are keyed by date, so it would probably be worth converting the provided date to the format used by the exceptions object as keys and seeing if it's there without the find.

Also, returning `boolean | undefined` might be easier, since undefined is what would be returned should the exception not exist.

**Review by chr1sjf0x — COMMENTED** [review]
This is certainly a read-ability improvement and simplification.

What do you think about keeping the `original` (aka prior to our changes) version of this method, and only executing your new version if there are exceptions?

File: src/addBusinessDays/index.ts [hunk 14]
```diff
@@ -37,110 +37,65 @@ export default function addBusinessDays(
...
-  let restDays = Math.abs(amount % businessDays.length)
-  // Loops over remaining days
-  while (restDays > 0) {
-    date.setDate(date.getDate() + sign)
-    if (!isNonWorkingDay(date)) restDays -= 1
+  // returns a boolean if the date has an exception
+  // true means the date is a working day
+  // false means the date is not a working day
+  const findException = (date: Date): boolean | undefined => {
+    return exceptions[format(date, 'MM/dd/yy')]
```
Comment (ksk5280) [inline src/addBusinessDays/index.ts:60]: @chr1sjf0x This now formats the date to avoid doing a `find`. However, this check won't work correctly if the exception is not in the right format. 

We could loop through each exception to check for whether the exception has the correct format, but would that somewhat defeat the purpose of this change? Other options that I'm not thinking of?

File: src/addBusinessDays/index.ts (hunk 12 above)
Comment (ksk5280) [reply src/addBusinessDays/index.ts]: I added a test to check for this, so if we refactor later and add the `businessDays.length` back in, that test will fail.

File: src/addBusinessDays/index.ts (hunk 14 above)
Comment (chr1sjf0x) [reply src/addBusinessDays/index.ts:60]: I'd be inclined to say having them in the correct format is a requirement for this to work correctly which we don't/shouldn't need to validate at this level. It is stated in the javadoc associated with the exceptions field after all 🤷

File: src/addBusinessDays/index.ts (hunk 12 above)
Comment (chr1sjf0x) [reply src/addBusinessDays/index.ts]: Great, TY 👍

File: src/subBusinessDays/test.ts [hunk 15]
```diff
@@ -88,32 +88,45 @@ describe('subBusinessDays', () => {
...
+
+    it('can account for businessDays and exception options', () => {
+      const result = subBusinessDays(new Date(2022, 0 /* Jan */, 24), 11, {
+        // given businessDays of Mon-Sat
+        businessDays: [1, 2, 3, 4, 5, 6],
         exceptions: {
-          '01/17/2022': false,
-          '01/10/2022': false,
+          '01/17/22': false,
+          '01/10/22': false,
```
Comment (chr1sjf0x) [inline src/subBusinessDays/test.ts:126]: Question: Is this the format of our exceptions in BP?

**Review by chr1sjf0x — APPROVED** [review]
TY @ksk5280

File: src/subBusinessDays/test.ts (hunk 15 above)
Comment (ksk5280) [reply src/subBusinessDays/test.ts:126]: yep 👍
